### PR TITLE
firmware repository resource added

### PR DESCRIPTION
**Pull request by AnikaAgiwal2711**
# Description
A few sentences describing the overall goals of the pull request's commits.

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| |

# Checklist:

- [X] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [ ] I have verified that new and existing unit tests pass locally with my changes
- [ ] I have not allowed coverage numbers to degenerate
- [ ] I have maintained at least 80% code coverage
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] Backward compatibility is not broken

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

### CODE REVIEW

File: docs/resources/firmware_repository.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
+---
+# Copyright (c) 2023-2024 Dell Inc., or its subsidiaries. All Rights Reserved.
```
Comment (Krunal-Thakkar) [inline docs/resources/firmware_repository.md:2]: 2024 only

File: powerflex/provider/firmware_repository_resource.go [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+	}
+	frState := helper.UpdateFrimwareRepositoryState(frDetails, state)
+	diags = resp.State.Set(ctx, frState)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+}
+
+// Function used to Update fault set Resource
```
Comment (shenda1) [inline powerflex/provider/firmware_repository_resource.go:183]: Firmware Repository

File: powerflex/provider/firmware_repository_resource.go [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+	if err != nil {
+		resp.Diagnostics.AddError(
+			fmt.Sprintf("Could not Upload the compliance File"),
+			err.Error(),
+		)
+		return
+	}
+
+	// will loop until the file gets loaded successfully and status becomes available
+	for true {
```
Comment (rounak-adhikary) [inline powerflex/provider/firmware_repository_resource.go:110]: This loop needs to have a timeout condition

File: powerflex/provider/firmware_repository_resource.go [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+	diags := req.Plan.Get(ctx, &plan)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	ucParam := &scaleiotypes.UploadComplianceParam{
+		SourceLocation: plan.SourceLocation.ValueString(),
+	}
+
+	if !plan.Username.IsUnknown() && !plan.Password.IsUnknown() {
```
Comment (shenda1) [inline powerflex/provider/firmware_repository_resource.go:94]: Username and Password are optional params. So I think IsUnknown() will not be applicable. Please check.

File: powerflex/provider/firmware_repository_resource.go [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+			return
+		}
+		if frDetails.State == "available" {
+			break
+		} else if frDetails.State == "needsApproval" {
+			if plan.Approve.ValueBool() {
+				err := r.gatewayClient.ApproveUnsignedFile(frDetails.ID)
+				if err != nil {
+					resp.Diagnostics.AddError(
+						fmt.Sprintf("Could not Upload the compliance File"),
```
Comment (rounak-adhikary) [inline powerflex/provider/firmware_repository_resource.go:127]:
```suggestion
						fmt.Sprintf("Could not approve the compliance File"),
```

File: powerflex/provider/firmware_repository_resource.go [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+			}
+		} else if frDetails.State == "errors" {
+			resp.Diagnostics.AddError(
+				fmt.Sprintf("Could not Upload the compliance File"),
+				fmt.Sprint("error while uploading compliance file"),
+			)
+			return
+		} else if frDetails.State == "copying" {
+			continue
+		}
```
Comment (rounak-adhikary) [inline powerflex/provider/firmware_repository_resource.go:143]: This else if block is redundant, I believe.

File: go.sum [hunk 7]
```diff
@@ -33,8 +33,8 @@ github.com/cyphar/filepath-securejoin v0.2.4/go.mod h1:aPGpWjXOXUn2NCNjFvBE6aRxG
 github.com/davecgh/go-spew v1.1.0/go.mod h1:J7Y8YcW2NihsgmVo/mv3lAwl/skON4iLHjSsI+c5H38=
 github.com/davecgh/go-spew v1.1.1 h1:vj9j/u1bqnvCEfJOwUhtlOARqs3+rkHYY13jYWTU97c=
 github.com/davecgh/go-spew v1.1.1/go.mod h1:J7Y8YcW2NihsgmVo/mv3lAwl/skON4iLHjSsI+c5H38=
-github.com/dell/goscaleio v1.14.1 h1:SCHGLoOBKxQZ8EodChOLoIghcyhepbO5MLPRd4YQZ5c=
-github.com/dell/goscaleio v1.14.1/go.mod h1:h7SCmReARG/szFWBMQGETGkZObknhS45lQipQbtdmJ8=
+github.com/dell/goscaleio v1.14.2-0.20240509092006-6fc2fbf5680e h1:2bRQayfEUpCI5dES0ofG5rpSPMZwHt1trd5PU9Iy4LY=
```
Comment (shenda1) [inline go.sum:36]: Please point to the latest goscaleio commit

File: powerflex/provider/firmware_repository_resource.go [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+				err := r.gatewayClient.ApproveUnsignedFile(frDetails.ID)
+				if err != nil {
+					resp.Diagnostics.AddError(
+						fmt.Sprintf("Could not Upload the compliance File"),
+						err.Error(),
+					)
+					return
+				}
+			} else {
+				break
```
Comment (rounak-adhikary) [inline powerflex/provider/firmware_repository_resource.go:133]: Shouldn't we throw some error (or warning) if user has **NOT** provided `approve = true` and the compliance goes into a `needsApproval` state?

File: docs/resources/firmware_repository.md (hunk 1 above)
Comment (AnikaAgiwal2711) [reply docs/resources/firmware_repository.md]: Done

File: powerflex/provider/firmware_repository_resource.go (hunk 2 above)
Comment (AnikaAgiwal2711) [reply powerflex/provider/firmware_repository_resource.go]: Done

File: powerflex/provider/firmware_repository_resource.go (hunk 3 above)
Comment (AnikaAgiwal2711) [reply powerflex/provider/firmware_repository_resource.go]: Done

File: powerflex/provider/firmware_repository_resource.go (hunk 5 above)
Comment (AnikaAgiwal2711) [reply powerflex/provider/firmware_repository_resource.go]: Done

File: go.sum (hunk 7 above)
Comment (AnikaAgiwal2711) [reply go.sum]: Done

File: powerflex/provider/firmware_repository_resource.go (hunk 8 above)
Comment (AnikaAgiwal2711) [reply powerflex/provider/firmware_repository_resource.go:138]: I don't think that we should add a warning.

File: powerflex/provider/firmware_repository_resource.go (hunk 4 above)
Comment (AnikaAgiwal2711) [reply powerflex/provider/firmware_repository_resource.go]: Done

File: powerflex/provider/firmware_repository_resource.go (hunk 8 above)
Comment (AnikaAgiwal2711) [reply powerflex/provider/firmware_repository_resource.go]: As per the discussion with Rounak, I've added the warning

**Review by rounak-adhikary — DISMISSED** [review]

**Review by shenda1 — DISMISSED** [review]

**Review by shenda1 — DISMISSED** [review]

**Review by Krunal-Thakkar — DISMISSED** [review]

**Review by rounak-adhikary — APPROVED** [review]

**Review by shenda1 — APPROVED** [review]

**Pull request merged**